syntax = "proto3";

package inventory;

service InventoryService {
  rpc AddProduct(Product) returns (ProductResponse) {}
  rpc GetProduct(ProductRequest) returns (Product) {}
  rpc UpdateStock(StockUpdateRequest) returns (Product) {}
  rpc RemoveProduct(ProductRequest) returns (RemoveResponse) {}
  rpc ListProducts(Empty) returns (ListProductsResponse) {}
  rpc StreamStockAlerts(AlertRequest) returns (stream StockAlert) {}
}

message Product {
  string id = 1;
  string name = 2;
  int32 quantity = 3;
}

message ProductRequest {
  string id = 1;
}

message StockUpdateRequest {
  string product_id = 1;
  int32 delta = 2;
}

message AlertRequest {
  int32 threshold = 1;
}

message StockAlert {
  string product_id = 1;
  string product_name = 2;
  int32 current_quantity = 3;
}

message ProductResponse {
  string id = 1;
  bool success = 2;
}

message RemoveResponse {
  bool success = 1;
}

message ListProductsResponse {
  repeated Product products = 1;
}

message Empty {}